dcr: spv will panic on retry in spv loop.

Here:
https://github.com/decred/dcrdex/blob/1ecfbc03e2c5d1359b5d51b32650e6cbff29b6b0/client/asset/dcr/spv.go#L430-L443

This was recently found while developing libwallet for cake, if the error has nothing to do with the context, this will panic on the second try of syncer.Run with closing a closed channel when `close(s.initialSyncDone)` happens for a second time. Unsure if intended on the dcrwallet side but doesn't seem to come up much anyway. The fix for libwallet is to recreate the syncer on every try. https://github.com/decred/libwallet/pull/18

Wow, actually hit this today on mainnet. Maybe it's due to some recent change:
```
2025-03-18 02:04:43.594 [INF] CORE: notify: |SUCCESS| (walletconfig) Wallet connectivity restored - Decred wallet has reestablished connectivity.
panic: close of closed channel

goroutine 19964016 [running]:
decred.org/dcrwallet/v4/spv.(*Syncer).Run.func12()
        /home/joe/go/pkg/mod/decred.org/dcrwallet/v4@v4.3.0/spv/sync.go:453 +0x2f8
golang.org/x/sync/errgroup.(*Group).Go.func1()
        /home/joe/go/pkg/mod/golang.org/x/sync@v0.11.0/errgroup/errgroup.go:78 +0x54
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 289
        /home/joe/go/pkg/mod/golang.org/x/sync@v0.11.0/errgroup/errgroup.go:75 +0x98
```

--- a/client/asset/dcr/spv.go
+++ b/client/asset/dcr/spv.go
@@ -153,8 +153,7 @@ type spvWallet struct {
 	dir               string
 	chainParams       *chaincfg.Params
 	log               dex.Logger
-	spv               spvSyncer // *spv.Syncer
-	bestSpvPeerHeight int32     // atomic
+	bestSpvPeerHeight int32 // atomic
 	tipChan           chan *block
 	gapLimit          uint32
 
@@ -164,6 +163,9 @@ type spvWallet struct {
 
 	cancel context.CancelFunc
 	wg     sync.WaitGroup
+
+	spvMtx sync.RWMutex
+	spv    spvSyncer // *spv.Syncer
 }
 
 var _ Wallet = (*spvWallet)(nil)
@@ -393,19 +395,10 @@ func (w *spvWallet) startWallet(ctx context.Context) error {
 	w.dcrWallet = &extendedWallet{dcrw}
 	w.db = db
 
-	var connectPeers []string
-	switch w.chainParams.Net {
-	case wire.SimNet:
-		connectPeers = []string{"localhost:19560"}
-	}
-
-	spv := newSpvSyncer(dcrw, w.dir, connectPeers)
-	w.spv = spv
-
 	w.wg.Add(2)
 	go func() {
 		defer w.wg.Done()
-		w.spvLoop(ctx, spv)
+		w.spvLoop(ctx)
 	}()
 	go func() {
 		defer w.wg.Done()
@@ -427,8 +420,11 @@ func (w *spvWallet) stop() {
 	w.log.Info("SPV wallet closed")
 }
 
-func (w *spvWallet) spvLoop(ctx context.Context, syncer *spv.Syncer) {
+func (w *spvWallet) spvLoop(ctx context.Context) {
 	for {
+		// Create a new syncer after every failure or dcrwallet will
+		// panic because of closing of a closed channel.
+		syncer := w.newSpvSyncer()
 		err := syncer.Run(ctx)
 		if ctx.Err() != nil {
 			return
@@ -741,6 +737,8 @@ func (w *spvWallet) SignRawTransaction(ctx context.Context, baseTx *wire.MsgTx) 
 // SendRawTransaction broadcasts the provided transaction to the Decred network.
 // Part of the Wallet interface.
 func (w *spvWallet) SendRawTransaction(ctx context.Context, tx *wire.MsgTx, allowHighFees bool) (*chainhash.Hash, error) {
+	w.spvMtx.RLock()
+	defer w.spvMtx.RUnlock()
 	// TODO: Conditional high fee check?
 	return w.PublishTransaction(ctx, tx, w.spv)
 }
@@ -829,7 +827,9 @@ func (w *spvWallet) GetBlock(ctx context.Context, blockHash *chainhash.Hash) (*w
 		return block, nil
 	}
 
+	w.spvMtx.RLock()
 	blocks, err := w.spv.Blocks(ctx, []*chainhash.Hash{blockHash})
+	w.spvMtx.RUnlock()
 	if err != nil {
 		return nil, err
 	}
@@ -983,7 +983,9 @@ func (w *spvWallet) SyncStatus(ctx context.Context) (*asset.SyncStatus, error) {
 	height = utils.Clamp(height, 0, targetHeight)
 	ss.Blocks = uint64(height)
 
+	w.spvMtx.RLock()
 	ss.Synced, _ = w.spv.Synced(ctx)
+	w.spvMtx.RUnlock()
 
 	if rescanHash, err := w.dcrWallet.RescanPoint(ctx); err != nil {
 		return nil, fmt.Errorf("error getting rescan point: %w", err)
@@ -1003,7 +1005,9 @@ func (w *spvWallet) SyncStatus(ctx context.Context) (*asset.SyncStatus, error) {
 // spv peers. If no peers are connected, the last observed max peer height is
 // returned.
 func (w *spvWallet) bestPeerInitialHeight() int32 {
+	w.spvMtx.RLock()
 	peers := w.spv.GetRemotePeers()
+	w.spvMtx.RUnlock()
 	if len(peers) == 0 {
 		return atomic.LoadInt32(&w.bestSpvPeerHeight)
 	}
@@ -1047,6 +1051,8 @@ func (w *spvWallet) newVSPClient(vspHost, vspPubKey string, log dex.Logger) (*vs
 
 // rescan performs a blocking rescan, sending updates on the channel.
 func (w *spvWallet) rescan(ctx context.Context, fromHeight int32, c chan wallet.RescanProgress) {
+	w.spvMtx.RLock()
+	defer w.spvMtx.RUnlock()
 	w.dcrWallet.RescanProgressFromHeight(ctx, w.spv, fromHeight, c)
 }
 
@@ -1085,7 +1091,9 @@ func (w *spvWallet) PurchaseTickets(ctx context.Context, n int, vspHost, vspPubK
 		}
 	}
 
+	w.spvMtx.RLock()
 	res, err := w.dcrWallet.PurchaseTickets(ctx, w.spv, req)
+	w.spvMtx.RUnlock()
 	if err != nil {
 		return nil, err
 	}
@@ -1443,6 +1451,8 @@ func (w *spvWallet) cachedBlock(blockHash *chainhash.Hash) *wire.MsgBlock {
 
 // PeerCount returns the count of currently connected peers.
 func (w *spvWallet) PeerCount(ctx context.Context) (uint32, error) {
+	w.spvMtx.RLock()
+	defer w.spvMtx.RUnlock()
 	return uint32(len(w.spv.GetRemotePeers())), nil
 }
 
@@ -1458,15 +1468,25 @@ func (w *spvWallet) cleanBlockCache() {
 	}
 }
 
-func newSpvSyncer(w *wallet.Wallet, netDir string, connectPeers []string) *spv.Syncer {
+func (w *spvWallet) newSpvSyncer() *spv.Syncer {
+	w.spvMtx.Lock()
+	defer w.spvMtx.Unlock()
+	ew := w.dcrWallet.(*extendedWallet)
+	dcrw := ew.Wallet
+	var connectPeers []string
+	switch w.chainParams.Net {
+	case wire.SimNet:
+		connectPeers = []string{"localhost:19560"}
+	}
 	addr := &net.TCPAddr{IP: net.ParseIP("::1"), Port: 0}
-	amgr := addrmgr.New(netDir, net.LookupIP)
-	lp := p2p.NewLocalPeer(w.ChainParams(), addr, amgr)
-	syncer := spv.NewSyncer(w, lp)
+	amgr := addrmgr.New(w.dir, net.LookupIP)
+	lp := p2p.NewLocalPeer(dcrw.ChainParams(), addr, amgr)
+	syncer := spv.NewSyncer(dcrw, lp)
 	if len(connectPeers) > 0 {
 		syncer.SetPersistentPeers(connectPeers)
 	}
-	w.SetNetworkBackend(syncer)
+	dcrw.SetNetworkBackend(syncer)
+	w.spv = syncer
 	return syncer
 }
 

--- a/client/asset/dcr/spv_mixing.go
+++ b/client/asset/dcr/spv_mixing.go
@@ -50,7 +50,9 @@ func (w *spvWallet) mix(ctx context.Context) {
 
 	// Don't perform any actions while transactions are not synced
 	// through the tip block.
+	w.spvMtx.RLock()
 	synced, _ := w.spv.Synced(ctx)
+	w.spvMtx.RUnlock()
 	if !synced {
 		w.log.Tracef("Skipping account mixing: transactions are not synced")
 		return

